src/Artifacts.jl

```diff
@@ -0,0 +1,589 @@
+module Artifacts
+
+import Base: get, SHA1
+import ..depots1
+import ..GitTools: tree_hash, set_readonly
+using ..BinaryPlatforms
+import ..TOML
+import ..Types: parse_toml
+import ..PlatformEngines: download_verify_unpack, probe_platform_engines!, package
+
+export create_artifact, artifact_exists, artifact_path, remove_artifact, verify_artifact,
+       artifact_meta, artifact_hash, bind_artifact, unbind_artifact, download_artifact,
+       find_artifact_toml, ensure_artifact_installed, @artifact_str, archive_artifact
+
+## Philosophy of Artifacts:
+#
+# - At the lowest level, artifacts are content-addressed chunks of data.  Similarly to
+#   how packages are identified by UUIDs but have human-readable mappings through the
+#   registry, artifacts are identified by their SHA1 git tree hashes, but have human-
+#   readable mappings through `Artifact.toml` files.
+#
+# - Artifacts do not _need_ to have names bound to them; they are first and foremost
+#   identified by their tree hash.  A name is merely a pointer to a hash, and the hash
+#   itself is what is used in all operations except the initial lookup.
+#
+# - A single name can be associated with multiple hashes, where one is chosen at runtime.
+#   The canonical example of this is platform-dependent artifacts.  At `Artifact.toml`
+#   parse-time one of the list of hashes will be chosen as the best match for the
+#   current environment (Julia version, OS, processor architecture, etc...), and note
+#   that it is acceptable for there to be multiple possible choices (e.g. an AVX2-
+#   optimized build and a generic x86_64 build that will both run on a modern processor),
+#   however one single hash will be stably chosen from the multiple possibilities.
+#
+# - Artifacts can be created on-machine through the `create()` functional API, and can
+#   also be downloaded as tarballs from external webservers by embedding information into
+#   the `Artifact.toml` file.  This can be done programmatically through the
+#   `bind_artifact()` function.
+
+
+# Levels of API:
+#
+# - Base level:
+#   - create_artifact(): given no information, simply creates an artifact through a user-provided
+#     callback and returns the hash of the resultant file.
+#
+# - Hash level:
+#   - artifact_exists(hash): returns true if the hash exists on disk, false otherwise
+#   - artifact_path(hash):   returns the path to a hash on disk, throws if it does not exist
```

Comment:
Just an idea, but could we return `nothing` if the file does not exist and then get rid of the `artifact_exist` function completely?

Reply:
We could; I kind of like being able to easily calculate where an artifact _would_ be though, and I don't like the type instability making things like `joinpath(artifact_path(hash), "bin", "foo")` dangerous.  I'd rather get a "file does not exist" error than a MethodError there.

Reply:
You could do `ispath(artifiact_path(hash))`, no? Does the implementation do any more than that?

Reply:
The point being that if `artifact_path(hash)` returns `nothing` for something that is not installed, then even `ispath(artifact_path(hash))` will throw a `MethodError` (since `ispath()` uses `joinpath()` internally).  If it instead always returns a string, even if the thing is not installed, you get a much more easily-understood `ENOENT`.

Reply:
That was my point: if `artifact_path` returns the path of an existent artifact or where the artifact would get unpacked for a non-existent artifact, then `ispath(artifact_path(hash))` does what `artifact_exists` does and seems almost as easy to write.

Reply:
Yes, it's a [very simple function](https://github.com/JuliaLang/Pkg.jl/blob/4cb4d58e18022a9eaca124b0e039efc1283adfbd/src/Artifacts.jl#L144-L150):

Reply:
It is now less simple, because we need to search multiple depots.